libnetdata/socket/socket.c

```diff
@@ -1010,6 +1010,44 @@ int accept_socket(int fd, int flags, char *client_ip, size_t ipsize, char *clien
             strncpyz(client_ip, "UNKNOWN", ipsize - 1);
             strncpyz(client_port, "UNKNOWN", portsize - 1);
         }
+        char client_host[NI_MAXHOST+1];
+        if(getnameinfo((struct sockaddr *)&sadr, addrlen, client_host, (socklen_t)ipsize, NULL, 0, NI_NAMEREQD | NI_NUMERICSERV) != 0)
```

Comment:
I don't see why you're calling `getnameinfo()` again, given that we're calling it a few lines up. Can you refactor the code so that we only have one call to it, or is there another reason for which we need two calls?

Reply:
If we call it once then we don't know if the response was an ip-address or a hostname. In the case that it is a hostname we need the ip-address as well: it propagates outside of the function and it means that numeric patterns in the config will still work.

Reply:
The way it's structured now, that's true. Hence my question, if you could think of a way to refactor the code to achieve the proper flow with one call. :)

Reply:
If the flow is to get both an IP address and a hostname then it is not possible with a single call to `getnameinfo()`.

Reply:
The case that I'm thinking about is when somebody puts something like: "allow connections from = 1.2.3.4 host.network" because they want to allow based on either the numeric address or the resolved DNS. That's what I would do as the reverse-dns might require a PTR record in someone else's server (e.g. an ISP) and the numeric entry acts as fall-back. If I refactor it to use a single call then it won't be possible to use a fallback.

Reply:
Refactored the code to remove the second call from code-paths where possible, only called twice now when needed to perform both pattern matches. Also deferred the DNS lookup until after the pattern-match.

Reply:
That's great, thank you! Here are my thoughts: `getnameinfo()` _without the flag_ returns hostname or, failing to resolve that, the IP. So you could call it _only once_, get whatever it returns (if anything), check whether it matches an IP or a hostname. If it's a hostname, you can do the forward lookup and get the IP. If it's an IP then you have your IP. One call, some more logic on your end.

Let me know if I'm missing something here @mfundul @amoss.

Reply:
This would break the existing functionality in the case that a numeric pattern is being used and reverse DNS is setup on the connecting IP. The single call would resolve e.g. 4.3.2.1.in-address.arpa. while the user may have specified 1.2.3.4 as their pattern. This is outside of user control if they do not own the ip block.

Reply:
That doesn't seem right. If a reverse DNS is setup on the connecting IP, it will resolve to whatever the PTR is, i.e. an FQDN. Then you'd validate that FQDN by doing a forward lookup and get the IP. Then you'd compare both the IP you got and the FQDN with your list. This, however, as per the `man` page is not wise, as someone could create a malicious PTR to spoof their origin. So we need the two calls (or, at the very least a call to `getaddrinfo()` and one to `getnameinfo()`, as per the man page.)

Reply:
Reverse DNS does not always have a PTR to a FQDN. Most ISPs will set it on addresses inside their block. So, for example, my home connection is very similar to ua-85-228-174-160.bbcust.telenor.se and this is what I see when I run "w" on a server that I'm logged into it. If I was using the existing functionality then I would have set:

`allow connections from = 85.228.174.160`

This pattern would not match the hostname returned by the single `getnameinfo()` when NI_NUMERICHOST is not set. We cannot perform an automatic mapping ourselves to determine this because there is no standard for the syntax of these psuedo hostnames (i.e. each ISP will format them slightly differently).

So to support both numeric IP addresses and hostnames (in a secure way to avoid the spoofing of PTRs) we end up needing all three calls -> two to `getaddrinfo() `to handle both numeric IPs and hostnames, and the final call to `getnameinfo()` if we match the hostname against a pattern.

Reply:
But `ua-85-228-174-160.bbcust.telenor.se` *is* an FQDN. And a forward lookup would point to 85.228.174.160. Adding both would match the `allow connections from` with the IP only. The only thing necessitating the multiple calls is the spoofing imho. But, then again, I've been working for many hours straight so I may be missing something here. :)

Reply:
True ;) I meant that it is not a unique (or intentional) hostname for my ip address.

If the user adds both into the `allow connections from` config then it would also work with a single `getaddrinfo()` call. But... there will be existing configurations that do not have both (because it was not needed previously) and these would suddenly break.

Also, the machine that I am connecting to netdata from may not have reverse DNS when I write the config, and it would work fine. But a change to a third party's DNS configuration can then break my access - outside of my control. Leaving both calls in the code avoids all of this and ensures that it can work with either numeric IPs or hostnames.

Reply:
Maybe I didn't do such a good job explaining myself, but I don't think any existing configurations (with IP only) would break, as long as the PTR records had matched A records, (as per the RFC.) The only case in which anything would break would be that of an unmatched PTR record (i.e. missing the A record) and/or spoofed PTR records. But this discussion is getting too academic, as we both agree we need the extra calls to address the spoofing anyway! Thanks Andrew :)